src/Data/HashMap/Linear.hs

```diff
@@ -63,32 +62,34 @@ import GHC.Stack
 
 -- | A mutable hashmap with a linear API
 data HashMap k v where
-  -- | HashMap (size,count) array-of-robin-values
-  -- The size is the amount of memory the array takes up.
+  -- |
   -- The count is the number of stored mappings.
-  -- Our sparseness invariant: count*3 <= size.
-  HashMap :: Size -> Count -> RobinArr k v #-> HashMap k v
+  -- Our sparseness invariant: count*3 <= size arr.
+  HashMap :: Count -> RobinArr k v #-> HashMap k v
 
 -- INVARIANTS:
 --   * Cells are empty iff the PSL is -1.
 --   * Each (RobinVal k v) has the correct PSL
 --   * We NEVER evaluate the key-val pair for PSL -1
 --     since it might be (error "some message")
 
--- | The size allocated for the array
-newtype Size = Size Int
-  deriving (Prelude.Num)
-
 -- | The number of pairs stored in the hashmap
 newtype Count = Count Int
   deriving (Prelude.Num)
 
 -- | An array of Robin values
-type RobinArr k v = Array (RobinVal k v)
+type RobinArr k v = Array (Maybe (RobinVal k v))
```

Comment:
If you're changing the representation you need to state the invariants, even if obvious:

> Each cell is `Nothing` if empty and is a `RobinVal` with the correct PSL otherwise.